lxd/device/nic_ovn.go

```diff
@@ -996,7 +996,7 @@ func (d *nicOVN) State() (*api.InstanceStateNetwork, error) {
 	// OVN only supports dynamic IP allocation if neither IPv4 or IPv6 are statically set.
 	if d.config["ipv4.address"] == "" && d.config["ipv6.address"] == "" {
 		instanceUUID := d.inst.LocalConfig()["volatile.uuid"]
-		dynamicIPs, err := d.network.InstanceDevicePortDynamicIPs(instanceUUID, d.name)
+		dynamicIPs, err := d.network.InstanceDevicePortIPs(instanceUUID, d.name)
 		if err == nil {
 			for _, dynamicIP := range dynamicIPs {
```

Comment:
logic here needs to exclude any ips that are specified in the static NIC config otherwise we'll end up with duplicates

Reply:
I was looking at this one, but this part of code is executed only if both `ipv4.address` and `ipv6.address` are empty (line 997), so there should be no need for that.

Am I missing something?

Reply:
Can you test if that statement: 

> // OVN only supports dynamic IP allocation if neither IPv4 or IPv6 are statically set.

Is still true with sticky IPs in play?

Reply:
Setting only one static IP (either IPv4 or IPv6) results in the IP of the other family to be dynamically set. So, this should no longer be true?

Reply:
Setting IPv4 to static also sets IPv6 to EU64 statically in OVN:

```
lxc launch ubuntu:jammy c1 --network ovn1 -d eth0,ipv4.address=10.143.147.10

sudo ovn-nbctl list logical_switch_port
addresses           : ["00:16:3e:8b:45:8f 10.143.147.10 fd42:875e:7708:3d62:216:3eff:fe8b:458f"]
```

But setting IPv6 statically without IPv4 results in another bug:

```
lxc launch ubuntu:jammy c1 --network ovn1 -d eth0,ipv6.address=fd42:875e:7708:3d62:216:3eff:fe8b:458f
Creating c1
Starting c1                                 
Error: Failed to start device "eth0": Failed setting up OVN port: Insufficient dynamic addresses allocated
Try `lxc info --show-log local:c1` for more info
```

Please can you look into that also.

Reply:
If `ipv6.dhcp.stateful` is enabled, setting static IPv4 results in IPv6 (EUI64) not being shown.
```go
// OVN only supports dynamic IP allocation if neither IPv4 or IPv6 are statically set
if d.config["ipv4.address"] == "" && d.config["ipv6.address"] == "" {
   ...
}
```

Because it is listed under leases:
```
$ lxc launch ubuntu:22.04 c2 -n ovn -d eth0,ipv4.address=10.100.100.15

$ lxc ls
+------+---------+----------------------+------------------------------------------------+-----------+-----------+
| NAME |  STATE  |         IPV4         |                      IPV6                      |   TYPE    | SNAPSHOTS |
+------+---------+----------------------+------------------------------------------------+-----------+-----------+
| c2   | RUNNING | 10.100.100.15 (eth0) |                                                | CONTAINER | 0         |
+------+---------+----------------------+------------------------------------------------+-----------+-----------+


$ lxc network list-leases ovn
+----------+-------------------+-----------------------------------------+---------+
| HOSTNAME |    MAC ADDRESS    |               IP ADDRESS                |  TYPE   |
+----------+-------------------+-----------------------------------------+---------+
| c2       | 00:16:3e:86:d5:d5 | 10.100.100.15                           | STATIC  |
+----------+-------------------+-----------------------------------------+---------+
| c2       | 00:16:3e:86:d5:d5 | fd42:1000:1000:1000:216:3eff:fe86:d5d5  | DYNAMIC |
+----------+-------------------+-----------------------------------------+---------+
| ovn.gw   |                   | 10.100.100.1                            | GATEWAY |
+----------+-------------------+-----------------------------------------+---------+
| ovn.gw   |                   | fd42:1000:1000:1000::1                  | GATEWAY |
+----------+-------------------+-----------------------------------------+---------+
```

However, if static IPv6 is set (and IPv4 is not), OVN router does not provide any dynamic IP (even though there is enough IPs available).

Commenting out the error line results in only static IPv6 being generated:
```
lxc launch ubuntu:22.04 c1 -n ovn -d eth0,ipv6.address=fd42:1000:1000:1000:1234:1234:1234:1234

$ lxc ls
+------+---------+----------------------+------------------------------------------------+-----------+-----------+
| NAME |  STATE  |         IPV4         |                      IPV6                      |   TYPE    | SNAPSHOTS |
+------+---------+----------------------+------------------------------------------------+-----------+-----------+
| c1   | RUNNING |                      | fd42:1000:1000:1000:1234:1234:1234:1234 (eth0) | CONTAINER | 0         |
+------+---------+----------------------+------------------------------------------------+-----------+-----------+

$ lxc network list-leases ovn
+----------+-------------------+-----------------------------------------+---------+
| HOSTNAME |    MAC ADDRESS    |               IP ADDRESS                |  TYPE   |
+----------+-------------------+-----------------------------------------+---------+
| c1       | 00:16:3e:06:31:3d | fd42:1000:1000:1000:1234:1234:1234:1234 | STATIC  |
+----------+-------------------+-----------------------------------------+---------+
| ovn.gw   |                   | 10.100.100.1                            | GATEWAY |
+----------+-------------------+-----------------------------------------+---------+
| ovn.gw   |                   | fd42:1000:1000:1000::1                  | GATEWAY |
+----------+-------------------+-----------------------------------------+---------+
```

---

This is actually the case for both IPv6 and IPv4, but when IPv4 is set, EUI64 is generated which prevents the error. 

I'll dive deeper into this after fixing other issues.